### PR TITLE
Allow GraphQL to filter on nested property

### LINKED ISSUE #1714: Search Filter does not filter collection by nested relation field in GraphQL
Hi!
I've been recently testing GraphQL endpoint provided by this amazing framework.
So I've encountered on follwing SearchFilter issue.

<details>
<summary>
I've following entities (simplified):
</summary>

```php
// src/Entity/Article.php

/**
 * @ApiResource(
 *     attributes={
 *         "filters": {"app.article.search_filter", "app.article.group_filter"},
 *         "normalization_context": {"groups": {"ArticleRead"}},
 *     },
 *     graphql={
 *         "query": {
 *             "filters": {"app.article.search_filter"},
 *             "normalization_context": {"groups": {"ArticleRead", "CategoryRead"}},
 *         },
 *     },
 * })
 * @Entity
 */
class Article
{
    /**
     * @ORM\Id
     * @ORM\Column(type="uuid")
     * @ORM\GeneratedValue(strategy="CUSTOM")
     * @ORM\CustomIdGenerator(class="Ramsey\Uuid\Doctrine\UuidGenerator")
     *
     * @Groups({"ArticleRead"})
     */
    protected $id;

    ...

    /**
     * @ORM\ManyToOne(targetEntity="Category")
     * @ORM\JoinColumn(name="category_id", referencedColumnName="id", onDelete="CASCADE", nullable=false)
     *
     * @Assert\NotBlank
     *
     * @Groups({"ArticleRead"})
     */
    protected $category;
}
```

```php
// src/Entity/Category.php

/**
 * @ApiResource
 * @Entity
 */
class Category
{
    /**
     * @ORM\Id
     * @ORM\Column(type="uuid")
     * @ORM\GeneratedValue(strategy="CUSTOM")
     * @ORM\CustomIdGenerator(class="Ramsey\Uuid\Doctrine\UuidGenerator")
     *
     * @Groups({"CategoryRead"})
     */
    protected $id;

    /**
     * @ORM\Column(type="string", unique=true)
     *
     * @Gedmo\Slug(fields={"name"}, separator="-", updatable=true, unique=true)
     *
     * @Groups({"CategoryRead"})
     */
    protected $code;

    /**
     * @ORM\Column(type="string", nullable=false)
     *
     * @Assert\NotBlank
     * @Assert\Length(min="3", max="100")
     *
     * @Groups({"CategoryRead"})
     */
    protected $name;
}
```
</details>

<details>
<summary>
and I've declared following filters (old way):
</summary>

```yaml
app.article.search_filter:
        parent: 'api_platform.doctrine.orm.search_filter'
        arguments: [ { category.code: 'exact' } ]
        tags: [ 'api_platform.filter' ]

app.article.group_filter:
        parent: 'api_platform.serializer.group_filter'
        arguments: [ 'group', false, ['CategoryRead'] ]
        tags: [ 'api_platform.filter' ]
```
</details>

Filter works fine using normal REST:
`GET /articles?category.code=news&group[]=CategoryRead`

but using GraphQL:
`GET /graphql?query={articles(category_code:"news"){edges{node{category{code}}}}}`

<details>
<summary>
It returns an unfiltered collection:
</summary>

```json
{
  "data": {
    "articles": {
      "edges": [
        {
          "node": {
            "category": {
              "code": "news"
            }
          }
        },
        {
          "node": {
            "category": {
              "code": "article"
            }
          }
        },
        {
          "node": {
            "category": {
              "code": "announcement"
            }
          }
        }
      ]
   }
}
```
</details>

<details>
<summary>
I've tried also new way via annotation:
</summary>

```php
// src/Entity/Article.php

/**
 * @ApiFilter(SearchFilter::class, properties={"category.code": "exact"})
 */
class Article {
  ...
}
```
</details>

but the behaviour is exactly the same, in REST context works in GraphQL it does not.
I've checked and filtering works for not nested fields. 

Stack:
- api-platform/core v2.2.1, v2.2.2
- symfony v4.0.4
- php 7.1/7.2

Here I provide my test api endpoints, with symfony profiler enabled, so you can see the details, hope it helps:

*REST*
https://knit-test-api.tk/articles?category.code=news&group%5B%5D=CategoryRead
*GraphQL*:
https://knit-test-api.tk/graphql?query={articles(category_code:%22news%22){edges{node{category{code}}}}}
*Travis CI build with feature tests*:
https://travis-ci.org/knit-pk/api-v1-php/jobs/343330287#L708
Git branch (full codebase):
*https://github.com/knit-pk/api-v1-php/tree/feature/graphql-filters-test*

Thanks for your time.
PS: I appreciate your work very much, and really enjoy developing an API with API Platform!

### PR DESCRIPTION
**Pull request by antograssiot**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1714, #1867
| License       | MIT
| Doc PR        | 

Possible replacement for #1743 without updating the filter but just the way filters are passed to the data provider.

ping @alanpoulain  @IonBazan

### CODE REVIEW

**Comment by IonBazan** [comment]
Please also add a test for order and date filter to make sure it's ok.

**Comment by antograssiot** [comment]
I will add it for date, ordering on nested property doesn't make much sense but I might be missing the point ?

**Comment by IonBazan** [comment]
Consider fetching posts ordered by author's rating (stars).

**Comment by alanpoulain** [comment]
I don't really like this way. The issue will be too obscure for an user if it used a custom filter with an underscore. I would prefer the other method.

**Comment by antograssiot** [comment]
Sure I figured out just after commenting. this patch will need more work because the keys of value would need to be cleaned when they are arrays and for the key `order`. I'll have a look tonight maybe.

**Comment by antograssiot** [comment]
So should I Spend more time on this to get also the nested filed ordering to work @alanpoulain or should I close it s the other PR is a prefered solution and should be extented for OrderFilter ?

**Comment by antograssiot** [comment]
I've added a naive workaround so ordering on nesting property value would work now.
What's your opinion ?

**Comment by IonBazan** [comment]
I wonder if it's possible to use nested property filters like:
`dummies(relatedDummy: {name: "foo"})` and `dummies(order: {relatedDummy: {name: "foo"}})` because in your solution, filters named: `related_name` and `related.name` are ambiguous.

**Comment by antograssiot** [comment]
@IonBazan I don't know much about the graphQL part yet, but at least the solution is the same for all filters used in graphQL context without changing the way it works for standard Rest API, that's why I went with that.
Whatever the solution will be at the end, it should be properly documented I think...

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php [hunk 1]
```diff
@@ -164,4 +167,22 @@ private function getSubresource(string $rootClass, array $rootResolvedFields, ar
             'collection' => $isCollection,
         ]);
     }
+
+    private function getNormalizedFilters($args)
+    {
+        $filters = $args;
+        foreach ($filters as $name => &$value) {
+            if ($name === $this->orderParameterName && \is_array($value)) {
```
Comment (alanpoulain) [inline src/GraphQl/Resolver/Factory/CollectionResolverFactory.php]: Why not generalizing it for all filters (i.e. only verifying if `$value` is an array)?

**Comment by antograssiot** [comment]
I do understand that for people using GraphQL only, it might be tricky to define filters using the dot notation and use and underscored notation when using it but I don't really see any other better solution for now...

File: features/graphql/filters.feature [hunk 2]
```diff
@@ -88,3 +88,51 @@ Feature: Collections filtering
...
     And the JSON node "data.dummies.edges[2].node.relatedDummies.edges[0].node.name" should be equal to "RelatedDummy13"
+
+  @createSchema
+  Scenario: Retrieve a collection filtered using the related search filter
+    Given there are 1 dummy objects having each 2 relatedDummies
+    And there are 1 dummy objects having each 3 relatedDummies
+    When I add "Accept" header equal to "application/hal+json"
+    And I send a "GET" request to "/dummies?relatedDummies.name=RelatedDummy31"
+    Then the response status code should be 200
+    And the response should be in JSON
```
Comment (alanpoulain) [inline features/graphql/filters.feature]: This test shouldn't be here.

File: features/graphql/filters.feature [hunk 3]
```diff
@@ -88,3 +88,51 @@ Feature: Collections filtering
...
+    And the response status code should be 200
+    And the JSON node "data.dummies.edges" should have 1 element
+
+  @createSchema
+  Scenario: Retrieve a collection ordered using nested properties
+    Given there are 2 dummy objects with relatedDummy
+    When I send the following GraphQL request:
+    """
+    {
+      dummies(order:{relatedDummy_name: "DESC"}) {
```
Comment (alanpoulain) [inline features/graphql/filters.feature]: Please add a space after `order:`.

File: src/GraphQl/Type/SchemaBuilder.php [hunk 4]
```diff
@@ -289,9 +291,12 @@ private function mergeFilterArgs(array $args, array $parsed, ResourceMetadata $r
         return $args;
     }
 
-    private function convertFilterArgsToTypes(array $args): array
+    private function convertFilterArgsToTypes(array $args, bool $normalizeKeys = false): array
     {
         foreach ($args as $key => $value) {
+            if ($normalizeKeys && (strpos($key, '.'))) {
```
Comment (alanpoulain) [inline src/GraphQl/Type/SchemaBuilder.php]: Same, why not generalizing it?

**Comment by alanpoulain** [comment]
Maybe we could use another character than the underscore to avoid confusion?

**Comment by IonBazan** [comment]
@alanpoulain I would go for `dummies?relatedDummy[anotherRelatedDummy][name]=foo` for REST and `dummies(relatedDummy: {anotherRelatedDummy: {name: "foo"}})` for GraphQL because it ensures consistency between the APIs but it is also a BC break.

**Comment by alanpoulain** [comment]
Seems fine to me. I think there is no BC break for now because GraphQL is treated as an experimental feature. It's not really clear enough for end-users though.

**Comment by alanpoulain** [comment]
I think I get it. You want to avoid characters replacement by having the same kind of notation for the REST part and the GraphQL one?

**Comment by IonBazan** [comment]
@alanpoulain Exactly, but I'm not sure if there are any drawbacks of this solution. This would require a total filters refactor and a lot of testing. I think the solution proposed by @antograssiot is good enough for now.

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php (hunk 1 above)
Comment (antograssiot) [reply src/GraphQl/Resolver/Factory/CollectionResolverFactory.php]: My first idea was that this is not necessary fo all filters (like DateFilter) for example and wanted to keep the modification as small as possible... but on a second thought, I don't see any issue in generalizing it.

Comment (alanpoulain) [reply src/GraphQl/Resolver/Factory/CollectionResolverFactory.php]: And it simplifies the code :wink:

File: features/graphql/filters.feature [hunk 5]
```diff
@@ -88,3 +88,50 @@ Feature: Collections filtering
     And the JSON node "data.dummies.edges[1].node.relatedDummies.edges" should have 0 elements
     And the JSON node "data.dummies.edges[2].node.relatedDummies.edges" should have 1 element
     And the JSON node "data.dummies.edges[2].node.relatedDummies.edges[0].node.name" should be equal to "RelatedDummy13"
+
+  @createSchema
+  Scenario: Retrieve a collection filtered using the related search filter
+    Given there are 1 dummy objects having each 2 relatedDummies
+    And there are 1 dummy objects having each 3 relatedDummies
+    When I add "Accept" header equal to "application/hal+json"
```
Comment (IonBazan) [inline features/graphql/filters.feature]: I think it shouldn't be here as we are testing GraphQL endpoint, not HAL.

**Comment by antograssiot** [comment]
all comments so far have been addressed.

File: src/GraphQl/Type/SchemaBuilder.php [hunk 6]
```diff
@@ -292,6 +292,9 @@ private function mergeFilterArgs(array $args, array $parsed, ResourceMetadata $r
     private function convertFilterArgsToTypes(array $args): array
     {
         foreach ($args as $key => $value) {
+            if (strpos($key, '.')) {
```
Comment (alanpoulain) [inline src/GraphQl/Type/SchemaBuilder.php]: Add a comment to explain?

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php [hunk 7]
```diff
@@ -164,4 +164,22 @@ private function getSubresource(string $rootClass, array $rootResolvedFields, ar
             'collection' => $isCollection,
         ]);
     }
+
+    private function getNormalizedFilters($args)
```
Comment (alanpoulain) [inline src/GraphQl/Resolver/Factory/CollectionResolverFactory.php]: Add ` :array` at the end and `array` for the parameter.

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php [hunk 8]
```diff
@@ -164,4 +164,22 @@ private function getSubresource(string $rootClass, array $rootResolvedFields, ar
...
+    {
+        $filters = $args;
+        foreach ($filters as $name => &$value) {
```
Comment (alanpoulain) [inline src/GraphQl/Resolver/Factory/CollectionResolverFactory.php]: Please don't use references.

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php [hunk 9]
```diff
@@ -164,4 +164,22 @@ private function getSubresource(string $rootClass, array $rootResolvedFields, ar
...
+            if (\is_array($value)) {
+                $filters[$name] = $this->getNormalizedFilters($value);
+                continue;
+            }
+
+            if (strpos($name, '_')) {
+                // Gives a chance to relations/nested fields
```
Comment (alanpoulain) [inline src/GraphQl/Resolver/Factory/CollectionResolverFactory.php]: Please add a final point.

**Comment by antograssiot** [comment]
done @alanpoulain

File: src/GraphQl/Type/SchemaBuilder.php [hunk 10]
```diff
@@ -292,6 +292,11 @@ private function mergeFilterArgs(array $args, array $parsed, ResourceMetadata $r
...
+                // Declare relations/nested fields in a GraphQL compatible syntax.
+                $args[str_replace('.', '_', $key)] = $value;
```
Comment (alanpoulain) [inline src/GraphQl/Type/SchemaBuilder.php]: Is this really working? Because we are not looping with the added value.

**Comment by dunglas** [comment]
Good one, thank you very much @antograssiot. Can you also open a doc PR? It looks important to have this mentioned in the documentation PR (especially for REST actually).

**Comment by antograssiot** [comment]
Sure, I'll prepare a doc update this weekend or next week @dunglas

**Pull request merged**